miniapps/solvers/bramble_pasciak.hpp

```diff
@@ -0,0 +1,180 @@
+   const Operator *iprec, *pprec;
+   void UpdateVectors();
+
+public:
+   BPCGSolver() { }
+   BPCGSolver(const Operator &ipc, const Operator &ppc) { pprec = &ppc; iprec = &ipc; }
+
+#ifdef MFEM_USE_MPI
+   BPCGSolver(MPI_Comm comm_) : IterativeSolver(comm_) { }
+   BPCGSolver(MPI_Comm comm_, const Operator &ipc, const Operator &ppc)
+      : IterativeSolver(comm_) { pprec = &ppc; iprec = &ipc; }
+#endif
+
+   virtual void SetOperator(const Operator &op)
+   { IterativeSolver::SetOperator(op); UpdateVectors(); }
+
+   virtual void SetPreconditioner(Solver &pc)
+   { if (Mpi::Root()) { MFEM_WARNING("SetPreconditioner does NO effect to BPCGSolver.\n"); } }
+
+   virtual void SetIncompletePreconditioner(const Operator &ipc)
+   { iprec = &ipc; }
+
+   virtual void SetParticularPreconditioner(const Operator &ppc)
+   { pprec = &ppc; }
+
+   virtual void Mult(const Vector &b, Vector &x) const;
+};
+
+/// Bramble-Pasciak Solver for Darcy equation.
+/** Bramble-Pasciak Solver for Darcy equation.
+ *  The basic idea is to precondition the mass matrix M with a s.p.d. matrix Q
+ *  such that M - Q remains s.p.d. Then we can transform the block operator into a
+ *  s.p.d. operator under a modified inner product.
+ *  In particular, this enable us to implement modified versions of CG iterations,
+ *  that rely on efficient applications of the required transformations.
+ *
+ *  We offer a mass preconditioner based on a rescalling of the diagonal of the
+ *  element mass matrices M_T.
+ *  We consider Q_T := alpha * lambda_min * D_T, where D_T := diag(M_T), and
+ *  lambda_min is the smallest eigenvalue of the following problem
+ *                M_T x = lambda * D_T x.
+ *  alpha is a parameter that is stricly between 0 and 1.
+ *
+ *  For more details, see:
+ *  1. Vassilevski, Multilevel Block Factorization Preconditioners (Appendix F.3),
+ *     Springer, 2008.
+ *  2. James H. Bramble and Joseph E. Pasciak.
+ *     A Preconditioning Technique for Indefinite Systems Resulting From Mixed
+ *     Approximations of Elliptic Problems. Mathematics of Computation, 50:1–17, 1988.
+ */
+class BramblePasciakSolver : public DarcySolver
+{
+   mutable bool use_bpcg;
+   std::unique_ptr<IterativeSolver> solver_;
+   BlockOperator *oop_, *ipc_;
+   ProductOperator *mop_;
+   SumOperator *map_;
+   ProductOperator *ppc_;
+   BlockDiagonalPreconditioner *cpc_;
+   std::unique_ptr<HypreParMatrix> M_;
+   std::unique_ptr<HypreParMatrix> B_;
+   std::unique_ptr<HypreParMatrix> Q_;
+   OperatorPtr M0_;
+   OperatorPtr M1_;
+   Array<int> ess_zero_dofs_;
+
+   void Init(HypreParMatrix &M, HypreParMatrix &B,
+             HypreParMatrix &Q,
+             Solver &M0, Solver &M1,
+             const BPSParameters &param);
+public:
+   /// System and mass preconditioner are constructed from bilinear forms
+   BramblePasciakSolver(
+      const std::shared_ptr<ParBilinearForm> &mVarf,
+      const std::shared_ptr<ParMixedBilinearForm> &bVarf,
+      const BPSParameters &param);
```

Comment:
I don't know if exposing `shared_ptr` is great in terms of interface design. I know the ownership semantics are messy (i.e. the bilinear forms may end up owning some of the resulting operators), but `shared_ptr` puts some limitations on the caller that might not be ideal.

Reply:
What would be a preferred alternative?

Reply:
Just pass the objects by reference (or potentially const reference if possible).

Do I understand correctly that `BramblePasciakSolver` does not share ownership of the forms?

The `DarcyProblem` class can either store the bilinear forms as non-pointer members, or as e.g. `unique_ptr`. `GetMform` and `GetBform` can return (const?) references to those objects.

(To expand on this, I think `shared_ptr` is very useful, but it can be a bit tricky to use as part of the public interface. It forces the caller to also be using `shared_ptr`, and so is "viral" in certain sense).